### PR TITLE
Fixing sprites' image map regression.

### PR DESCRIPTION
**Pull request by aghassemi**
@Hixie

### CODE REVIEW

**Comment by abarth** [comment]
LGTM

**Comment by eseidelGoogle** [comment]
Do we have technology to test this sort of thing?

**Comment by aghassemi** [comment]
@eseidel Should be possible. Working on it.

**Comment by aghassemi** [comment]
Added tests for ImageMap that would have caught the regression. PTAL.

What I added here also implicitly tests the asset loading and image decoding stacks, which might be too much for what we need to test here, but without a way to inject a mock implementation of some interfaces in the middle of the asset loading stack (namely AssetBundleImageProvider), I am not sure if there a way to avoid this.

**Comment by eseidelGoogle** [comment]
Thank you very much for pushing our testing frontier forward. :)

File: packages/flutter_sprites/test/image_map_test.dart [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        new Uint8List.fromList(BASE64.decode(kTestImageBase64));
+    dataPipe.producer.write(bytes.buffer.asByteData());
+    dataPipe.producer.handle.close();
+    return new Future<mojo.MojoDataPipeConsumer>.value(dataPipe.consumer);
+  }
+
+  @override
+  Future<String> loadString(String key, {bool cache: true}) {
+    if (key == 'AssetManifest.json')
+      return new SynchronousFuture<String>(kTestManifest);
```
Comment (abarth) [inline packages/flutter_sprites/test/image_map_test.dart]: SynchronousFuture is very subtle to use properly.  It looks like the non-test loadString returns a normal Future.  We should probably do that here too.

File: packages/flutter_sprites/test/image_map_test.dart [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+const String kTestImageBase64 =
+    'iVBORw0KGgoAAAANSUhEUgAAAAEAAAABAQMAAAAl21bKAAAAA1BMVEUAAACnej3aAAAAAXRSTlMAQObYZgAAAApJREFUCNdjYAAAAAIAAeIhvDMAAAAASUVORK5CYII=';
+
+class TestAssetBundle extends CachingAssetBundle {
+  @override
+  Future<mojo.MojoDataPipeConsumer> load(String key) {
+    mojo.MojoDataPipe dataPipe = new mojo.MojoDataPipe();
+    final Uint8List bytes =
+        new Uint8List.fromList(BASE64.decode(kTestImageBase64));
+    dataPipe.producer.write(bytes.buffer.asByteData());
```
Comment (abarth) [inline packages/flutter_sprites/test/image_map_test.dart]: We should check the return value of this write call to ensure that the data pipe accepted all this data.

**Comment by abarth** [comment]
Nice!  LGTM

File: packages/flutter_sprites/test/image_map_test.dart (hunk 1 above)
Comment (aghassemi) [inline packages/flutter_sprites/test/image_map_test.dart]: done

File: packages/flutter_sprites/test/image_map_test.dart (hunk 2 above)
Comment (aghassemi) [inline packages/flutter_sprites/test/image_map_test.dart]: done

**Pull request merged**